Fix issue #1397 custom post types not queryable with slug

chrisAlwinYS opened the pull request:
### Your checklist for this pull request
Thanks for sending a pull request! Please make sure you click the link above to view the contribution guidelines, then fill out the blanks below.

🚨Please review the [guidelines for contributing](.github/CONTRIBUTING.md) to this repository.

- [ ] Make sure you are making a pull request against the **develop branch** (left side). Also you should start *your branch* off *our develop*.
- [ ] Make sure you are requesting to pull request from a **topic/feature/bugfix branch** (right side). Don't pull request from your master!

What does this implement/fix? Explain your changes.
---------------------------------------------------
Fix custom post types returning null when queried with SLUG idType  


Does this close any currently open issues?
------------------------------------------
#1397 


Any relevant logs, error output, GraphiQL screenshots, etc?
-------------------------------------
![Screen Shot 2020-08-08 at 2 56 21 PM](https://user-images.githubusercontent.com/49703444/89701005-577f6300-d987-11ea-8685-0db07f1eb372.png)


Any other comments?
-------------------
…


Where has this been tested?
---------------------------
**Operating System:** MacOS 10.15.6

**WordPress Version:** 5.4.2

jasonbahl commented:
@chrisAlwinYS Thanks for this PR! Much needed fix. 🙏 

I'm running into some funky issues with this fix implemented. It's actually returning the `post_type_object` for me instead of the `post`. Possibly I have something else I've done locally to mess with things 🤔 .

I'll make sure I get to a clean state and check things out again. Hopefully have something more to report here shortly.

chrisAlwinYS commented:
@jasonbahl No worries. Happy to help  

That's an odd outcome, are your query vars in the NodeResolver looking similar to: `Array
(
    [page] => 
    [name] => test-title
    [post_type] => blog
)
`

jasonbahl commented:
@chrisAlwinYS so when I step through things with your change in place, and then add a `wp_send_json( $this->wp );` on line 284 of NodeResolver.php, I get the following output:

![Screen Shot 2020-08-17 at 3 56 08 PM](https://user-images.githubusercontent.com/1260765/90448317-5a263a80-e0a2-11ea-8094-05efecc7d4d7.png)

I think we need the following:

```php
case 'slug':
  return $context->node_resolver->resolve_uri( $args['id'], [ 'name' => $args['id'], 'post_type' => $post_type_object->name] );
  break;
case 'uri':
  return $context->node_resolver->resolve_uri( $args['id'], ['post_type' => $post_type_object->name] );
  break;
```

chrisAlwinYS commented:
@jasonbahl That's odd, I get matched rules and matched query. 
![Screen Shot 2020-08-18 at 10 51 37 AM](https://user-images.githubusercontent.com/49703444/90451871-ee46df00-e140-11ea-9391-8bce029db03d.png)

I have tested with your code, adding the name field and I am seeing no issues with pages, posts and custom post types. Happy to update the pull request if that's easier?

chrisAlwinYS commented:
@jasonbahl Yeah that's probably the difference. I see you've added the change. I haven't written tests using codeception before. I am happy to give it a go using the existing ones as a template.

seagyn commented:
Is there any hold up on this? Happy to help get it across the line as it's a massive blocker for me.

chrisAlwinYS commented:
I believe it's still waiting for review. Unless these tests need fixing but assumed as i have no control over them they are for reviewers/moderators. 

@jasonbahl I have updated the code to fix if a page has the same slug as a root of a custom post type, the page will return incorrectly. I discovered that before in 10.2 it was working but since is not. I have also added tests to check for this functionality.

jasonbahl commented on tests/wpunit/NodeBySlugTest.php:
```diff
@@ -167,4 +176,80 @@ public function testCustomPostBySlug() {
         $this->assertSame( $post->post_title, $actual['data']['customType']['title'] );
     }
+
+    function testPageWithNameOfPostType () {
```
@chrisAlwinYS ooh. Good catch here!

jasonbahl commented:
@chrisAlwinYS thanks for working on this! Apologies for the delay in reviewing.
